source/client.c

```diff
@@ -80,96 +101,106 @@ static void s_mqtt_client_shutdown(
-
-        aws_mutex_lock(&connection->outstanding_requests.mutex);
-
-        /* Successfully shutdown, so clear the outstanding requests */
-        aws_hash_table_clear(&connection->outstanding_requests.table);
-
-        aws_mutex_unlock(&connection->outstanding_requests.mutex);
-
-        MQTT_CLIENT_CALL_CALLBACK(connection, on_disconnect);
-
-    } else if (connection->state == AWS_MQTT_CLIENT_STATE_CONNECTING) {
-
-        AWS_LOGF_TRACE(
-            AWS_LS_MQTT_CLIENT, "id=%p: Initial connection attempt failed, calling callback", (void *)connection);
-
-        connection->state = AWS_MQTT_CLIENT_STATE_DISCONNECTED;
-
-        MQTT_CLIENT_CALL_CALLBACK_ARGS(connection, on_connection_complete, error_code, 0, false);
-
-    } else {
-
-        AWS_ASSERT(
-            connection->state == AWS_MQTT_CLIENT_STATE_CONNECTED ||
-            connection->state == AWS_MQTT_CLIENT_STATE_RECONNECTING ||
-            connection->state == AWS_MQTT_CLIENT_STATE_DISCONNECTED);
-
-        if (connection->state == AWS_MQTT_CLIENT_STATE_CONNECTED) {
+            AWS_LOGF_TRACE(AWS_LS_MQTT_CLIENT, "id=%p: Reconnect failed, retrying", (void *)connection);
 
+            aws_event_loop_schedule_task_future(
+                el, &connection->reconnect_task->task, connection->reconnect_timeouts.next_attempt);
+            break;
+        }
+        case AWS_MQTT_CLIENT_STATE_DISCONNECTING:
             AWS_LOGF_DEBUG(
                 AWS_LS_MQTT_CLIENT,
-                "id=%p: Connection lost, calling callback and attempting reconnect",
+                "id=%p: Disconnect completed, clearing request queue and calling callback",
                 (void *)connection);
 
-            connection->state = AWS_MQTT_CLIENT_STATE_RECONNECTING;
-            MQTT_CLIENT_CALL_CALLBACK_ARGS(connection, on_interrupted, error_code);
-        }
-
-        AWS_ASSERT(
-            connection->state == AWS_MQTT_CLIENT_STATE_RECONNECTING ||
-            connection->state == AWS_MQTT_CLIENT_STATE_DISCONNECTING ||
-            connection->state == AWS_MQTT_CLIENT_STATE_DISCONNECTED);
-
-        /* This will only be true if the user called disconnect from the on_interrupted callback */
-        if (connection->state == AWS_MQTT_CLIENT_STATE_DISCONNECTING) {
-
+            MQTT_CLIENT_CALL_CALLBACK(connection, on_disconnect);
+            break;
+        case AWS_MQTT_CLIENT_STATE_CONNECTING:
             AWS_LOGF_TRACE(
+                AWS_LS_MQTT_CLIENT, "id=%p: Initial connection attempt failed, calling callback", (void *)connection);
+            MQTT_CLIENT_CALL_CALLBACK_ARGS(connection, on_connection_complete, error_code, 0, false);
+            break;
+        case AWS_MQTT_CLIENT_STATE_CONNECTED: {
+            AWS_LOGF_DEBUG(
                 AWS_LS_MQTT_CLIENT,
-                "id=%p: Caller requested disconnect from on_interrupted callback, aborting reconnect",
+                "id=%p: Connection lost, calling callback and attempting reconnect",
```

Comment:
Connection interrupted (we should try to be consistent so users understand that these are related)